### PR TITLE
Fixes dispatching of derived type arguments to hub methods

### PR DESCRIPTION
**Pull request by NickDarvey**
- Uses `IsAssignableFrom` instead of equality comparison for matching invocation arguments against method parameters
- Adds test to validate behaviour

Addresses #11495

### CODE REVIEW

File: src/SignalR/server/SignalR/test/Internal/DefaultHubDispatcherTests.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.Extensions.Options;
+using Xunit;
+
+namespace Microsoft.AspNetCore.SignalR.Tests.Internal
+{
+    public class DefaultHubDispatcherTests
+    {
+        private class MockHubConnectionContext<TValue> : HubConnectionContext
```
Comment (NickDarvey) [inline src/SignalR/server/SignalR/test/Internal/DefaultHubDispatcherTests.cs]: The simplest way I could think to validate the values that get dispatched to the hub was via the HubConnectionContext, capturing the responses it receives. (The behaviour is buried a little deep inside the dispatcher.)

File: src/SignalR/server/SignalR/test/Internal/DefaultHubDispatcherTests.cs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+                    }
+                }
+
+                else throw new NotImplementedException();
+
+                return default;
+            }
+        }
+
+        private static DefaultHubDispatcher<THub> CreateDispatcher<THub>() where THub : Hub
```
Comment (NickDarvey) [inline src/SignalR/server/SignalR/test/Internal/DefaultHubDispatcherTests.cs]: I based the instantiation of our target (the dispatcher) on the way DefaultHubDispatcherBenchmark does it. I believe this is the minimum we require to put it under test.

Comment (anurse) [reply src/SignalR/server/SignalR/test/Internal/DefaultHubDispatcherTests.cs]: We have tests that do method dispatching in [HubConnectionHandlerTests.cs](https://github.com/aspnet/AspNetCore/blob/master/src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs) using a fake connection. I'd think these kinds of tests can go in that.

Comment (NickDarvey) [reply src/SignalR/server/SignalR/test/Internal/DefaultHubDispatcherTests.cs]: I've moved this test into HubConnectionHandlerTests and copied the patterns used in the other tests of dispatching for streaming hubs.

I'm referring to this test case as supporting 'derived parameters' in hub methods, so I've prefixed the related artefacts (classes, interfaces, methods) with 'DerivedParameter'.
If there's another way you'd like to refer to this, let me know and I'll update the names.

Comment (anurse) [reply src/SignalR/server/SignalR/test/Internal/DefaultHubDispatcherTests.cs]: Very nice! Thanks for doing that!

I think 'DerivedParameter' is a fine name. I'll let the team give a review and expect we'll be able to get this merged in.

**Review by BrennanConroy — APPROVED** [review]
Looks good, however I would like to avoid the usage of `TypeNameHandling.All`

File: src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs [hunk 3]
```diff
@@ -3576,6 +3576,62 @@ public async Task CanPassStreamingParameterToStreamHubMethod()
...
+            {
+                var argument = new StreamingHub.DerivedParameterTestObject { Value = "test" };
+                var protocolOptions = new NewtonsoftJsonHubProtocolOptions
+                {
+                    PayloadSerializerSettings = new JsonSerializerSettings()
+                    {
+                        // The usage of TypeNameHandling.All is a security risk.
+                        // If you're implementing this in your own application instead use your own 'type' field and a custom JsonConverter
+                        // or ensure you're restricting to only known types with a custom SerializationBinder.
+                        // See https://github.com/aspnet/AspNetCore/issues/11495#issuecomment-505047422
```
[the remark this answers is not in the document]
Comment (anurse) [reply src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs]: Agreed, we have a pretty hard policy about not using this kind of stuff, even in tests. I'd be OK with the [KnownTypesBinder](https://www.newtonsoft.com/json/help/html/SerializeSerializationBinder.htm) approach here.

[the remark this answers is not in the document]
Comment (NickDarvey) [reply src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs]: I've added a `DerivedParameterKnownTypesBinder` which limits the types to only the test type.

File: src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs [hunk 4]
```diff
@@ -3576,6 +3576,59 @@ public async Task CanPassStreamingParameterToStreamHubMethod()
...
+        {
+            using (StartVerifiableLog())
+            {
+                var argument = new StreamingHub.DerivedParameterTestObject { Value = "test" };
+                var protocolOptions = new NewtonsoftJsonHubProtocolOptions
+                {
+                    PayloadSerializerSettings = new JsonSerializerSettings()
+                    {
+                        TypeNameHandling = TypeNameHandling.All,
+                        SerializationBinder = StreamingHub.DerivedParameterKnownTypesBinder.Instance
```
Comment (halter73) [inline src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs]: I think we should keep the comment about TypeNameHandling.All being a potential security risk. We can just mention it's mitigated by providing a custom ISerializationBinder like you do here.

**Review by halter73 — APPROVED** [review]

File: src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs (hunk 4 above)
Comment (NickDarvey) [reply src/SignalR/server/SignalR/test/HubConnectionHandlerTests.cs]: Done

**Comment by anurse** [comment]
Thanks @NickDarvey !

**Pull request merged**